src/v/kafka/server/quota_manager.cc

```diff
@@ -35,6 +37,40 @@ using namespace std::chrono_literals;
 
 namespace kafka {
 
+namespace {
+constexpr size_t
+calculate_key(client_quota_rule rule, client_quota_type quota_type) {
+    // Assume the enums values are in sequence: [0, all_*.size())
+    static_assert(static_cast<size_t>(all_client_quota_rules[0]) == 0);
+    static_assert(static_cast<size_t>(all_client_quota_types[0]) == 0);
+    return static_cast<size_t>(rule) * all_client_quota_types.size()
+           + static_cast<size_t>(quota_type);
+}
+
+} // namespace
+
+void client_quotas_probe::record_throttle_time(
+  client_quota_rule rule, client_quota_type qt, std::chrono::microseconds t) {
+    _granular_metrics[calculate_key(rule, qt)].throttle_time.record(t.count());
+}
+
+ss::metrics::histogram client_quotas_probe::get_throttle_time(
+  client_quota_rule rule, client_quota_type qt) const {
+    return _granular_metrics[calculate_key(rule, qt)]
+      .throttle_time.internal_histogram_logform();
+}
+
+void client_quotas_probe::record_throughput(
+  client_quota_rule rule, client_quota_type qt, uint64_t count) {
+    _granular_metrics[calculate_key(rule, qt)].throughput.record(count);
+}
+
+ss::metrics::histogram client_quotas_probe::get_throughput(
+  client_quota_rule rule, client_quota_type qt) const {
+    return _granular_metrics[calculate_key(rule, qt)]
+      .throughput.internal_histogram_logform();
```

Comment:
I think it might be worth having a custom histogram here, we need 1ms to 30s, which is 15 buckets instead of 26, and I'd use fractional seconds.

Reply:
Are there any drawbacks to introducing a custom histogram here? Can customers easily create a diagram out of them? It looks like this would be the first custom histogram that's a public metric.

I went ahead and created a custom histogram for this assuming that it's fine to expose as a public metric.

Reply:
> Are there any drawbacks to introducing a custom histogram here?

I don't believe so

> Can customers easily create a diagram out of them?

I don't see why not, they're self-describing

> It looks like this would be the first custom histogram that's a public metric.

Well, there's a "custom histogram" for most latency measurements on the public_metrics, which has fewer buckets and is based on seconds.

> I went ahead and created a custom histogram for this assuming that it's fine to expose as a public metric.

:+1: